pkg/promlib/converter/prom.go

```diff
@@ -39,7 +39,7 @@ func ReadPrometheusStyleResult(jIter *jsoniter.Iterator, opt Options) backend.Da
 l1Fields:
 	for l1Field, err := iter.ReadObject(); ; l1Field, err = iter.ReadObject() {
 		if err != nil {
-			return rspErr(err)
+			return rspErr(fmt.Errorf("response from prometheus couldn't parsed. it is non-json"))
```

Comment:
Couldn't it be useful to return the actual error body, if it's non-JSON? That's what I suggested.

Reply:
That's the exact thing it was returning. I thought we didn't want that to return. 
In this case, the parser was working as expected. I am gonna close this PR and the issue, if you agree.

Reply:
It doesn't return the response body though, it returns a cryptic JSON decoding error like the following:

```
ReadObject: expect { or , or } or n, but found u, error found in #1 byte of ...|unexpected |..., bigger context ...|unexpected error |...
(Trace ID: 0000e63fb664a14cd0fdcc070c6d9b78) 
```

Reply:
So the message here
```
ReadObject: expect { or , or } or n, but found u, error found in #1 byte of ...|unexpected |..., bigger context ...|unexpected error |...
(Trace ID: 0000e63fb664a14cd0fdcc070c6d9b78) 
```
is what json parser gives us as error. We don't want that. I got it. 
We also don't want to return a custom error message. 
Then what exactly do we want to return? I know you said the response body, so if that's the whole authentication page in HTML, do we want to return that?

Reply:
I'm not a Grafana developer (any more), but to me the following seems sensible when non-JSON is returned:

1. Determine whether there's an error or not based on the HTTP response status code
2. Form the eventual error message from the response body (either the whole thing, or truncated, I'll leave to you to decide)

Reply:
I think what I suggest above corresponds to what is requested in the issue FWIW:

> [...] and parse the error as what it is.

Reply:
@itsmylife use the `res.body` as the error. This is what is being returned and it is not json so we completely miss what it is when we parse it.

I would do something like this

```
// pass the res.body into the converter
r := converter.ReadPrometheusStyleResult(iter, converter.Options{Dataplane: true}, res.Body)

// in the function ReadPrometheusStyleResult return the res.body as the error when json converting fails
for l1Field, err := iter.ReadObject(); ; l1Field, err = iter.ReadObject() {
		if err != nil {
			return rspErr(fmt.Errorf(res.body))
```

Reply:
I wouldn't use the `res.Body` because it might be a big HTML data of the authentication page. I don't think it is wise to return back that big HTML file. 

So I am returning a custom error message with the `err` itself. 
I won't return the body or a part of it. Because the `err` contains it already.

For instance, for this `res.Body`: `<!DOCTYPE html><html lang="en-US"><body><div>Authentication error</div></body></html>`
We got `err` as `ReadObject: expect { or , or } or n, but found <, error found in #1 byte of ...|<!DOCTYPE h|..., bigger context ...|<!DOCTYPE html><html lang="en-US"><body><div>Authen|...`

So it is there as truncated.